src/main/java/org/snomed/snowstorm/core/data/services/ConceptUpdateHelper.java

```diff
@@ -296,59 +297,94 @@ private void validateConcepts(Collection<Concept> concepts) {
-				// Associations for this type on both sides
-				Set<String> existingAssociationsOfType = existingAssociations.get(associationName);
-				Set<String> newAssociationsOfType = newAssociations.get(associationName);
-				// Iterate existing set
-				for (String existingAssociationTarget : existingAssociationsOfType) {
-					// If new set doesn't exist or doesn't contain existing association make it inactive.
-					if (newAssociationsOfType == null || !newAssociationsOfType.contains(existingAssociationTarget)) {
-						// Existing association should be made inactive
-						// Find the refset member for this association concept id and target concept id
-						String associationRefsetId = Concepts.historicalAssociationNames.inverse().get(associationName);
-						for (ReferenceSetMember existingMember : Optional.ofNullable(existingComponent.getAssociationTargetMembers()).orElse(Collections.emptySet())) {
-							if (existingMember.isActive() && existingMember.getRefsetId().equals(associationRefsetId)
-									&& existingAssociationTarget.equals(existingMember.getAdditionalField(ReferenceSetMember.AssociationFields.TARGET_COMP_ID))) {
-								existingMember.setActive(false);
-								existingMember.markChanged();
-								refsetMembersToPersist.add(existingMember);
-							}
-						}
-					}
+		Map<String, Set<String>> newVersionAssociations = newComponentVersion.getAssociationTargets();
+		if (newVersionAssociations == null) {
+			newVersionAssociations = new HashMap<>();
+		}
+		final Set<ReferenceSetMember> newVersionMembers = newComponentVersion.getAssociationTargetMembers();
+		final Set<ReferenceSetMember> existingVersionMembers = existingComponentVersion != null ? existingComponentVersion.getAssociationTargetMembers() : Collections.emptySet();
+
+		// New component version doesn't have refset members joined, it may have come from the REST API
+		// Attempt to match existing members to the association target key/value map
+		// Inactivate members which are no longer needed
+		// Create members which are in the key/value map but do not yet exist
+		Set<ReferenceSetMember> membersToKeep = new HashSet<>();
+		Set<ReferenceSetMember> membersToCreate = new HashSet<>();
+		Set<ReferenceSetMember> membersToRetire = new HashSet<>();
+		for (String associationTypeName : newVersionAssociations.keySet()) {
+			String associationRefsetId = Concepts.historicalAssociationNames.inverse().get(associationTypeName);
+			for (String associationValue : newVersionAssociations.get(associationTypeName)) {
+
+				ReferenceSetMember bestMember = getBestRefsetMember(associationRefsetId, ReferenceSetMember.AssociationFields.TARGET_COMP_ID, associationValue, newVersionMembers, existingVersionMembers);
+				if (bestMember != null) {
+					// Keep
+					membersToKeep.add(bestMember);
+				} else {
+					// Create new
+					bestMember = new ReferenceSetMember(newComponentVersion.getModuleId(), associationRefsetId, newComponentVersion.getId());
```

Comment:
Calling this constructor sets the active state to true implicitly already

Reply:
Yes, good point.